iXBRLViewerPlugin/iXBRLViewer.py

```diff
@@ -195,100 +198,145 @@ def getRelationships(self):
+            for qn in qnEnums:
+                self.addConcept(self.dts.qnameConcepts.get(qn))
+        else:
+            factData["v"] = f.value 
+            if f.value == INVALIDixVALUE:
+                factData["err"] = 'INVALID_IX_VALUE'
+
+        if f.format is not None:
+            factData["f"] = str(f.format)
+
+        if f.isNumeric:
+            if f.unit is not None and len(f.unit.measures[0]):
+                # XXX does not support complex units
+                unit = self.nsmap.qname(f.unit.measures[0][0])
+                aspects["u"] = unit
+            else:
+                # The presence of the unit aspect is used by the viewer to
+                # identify numeric facts.  If the fact has no unit (invalid
+                # XBRL, but we want to support it for draft documents),
+                # include the unit aspect with a null value.
+                aspects["u"] = None
+            d = inferredDecimals(f)
+            if d != float("INF") and not math.isnan(d):
+                factData["d"] = d
+
+        for d, v in f.context.qnameDims.items():
+            if v.memberQname is not None:
+                aspects[self.nsmap.qname(v.dimensionQname)] = self.nsmap.qname(v.memberQname)
+                self.addConcept(v.member)
+                self.addConcept(v.dimension, dimensionType = "e")
+            elif v.typedMember is not None:
+                aspects[self.nsmap.qname(v.dimensionQname)] = v.typedMember.text
+                self.addConcept(v.dimension, dimensionType = "t")
+
+        if f.context.isForeverPeriod:
+            aspects["p"] = "f"
+        elif f.context.isInstantPeriod and f.context.instantDatetime is not None:
+            aspects["p"] = self.dateFormat(f.context.instantDatetime.isoformat())
+        elif f.context.isStartEndPeriod and f.context.startDatetime is not None and f.context.endDatetime is not None:
+            aspects["p"] = "%s/%s" % (
+                self.dateFormat(f.context.startDatetime.isoformat()),
+                self.dateFormat(f.context.endDatetime.isoformat())
+            )
+
+        frels = self.footnoteRelationshipSet.fromModelObject(f)
+        if frels:
+            for frel in frels:
+                if frel.toModelObject is not None:
+                    factData.setdefault("fn", []).append(frel.toModelObject.id)
+
+        self.taxonomyData["facts"][f.id] = factData
+        self.addConcept(f.concept)
+
+    def addViewerToXMLDocument(self, xmlDocument, scriptUrl):
+        taxonomyDataJSON = self.escapeJSONForScriptTag(json.dumps(self.taxonomyData, indent=1, allow_nan=False))
+
+        for child in xmlDocument.getroot():
+            if child.tag == '{http://www.w3.org/1999/xhtml}body':
+                child.append(etree.Comment("BEGIN IXBRL VIEWER EXTENSIONS"))
+
+                e = etree.fromstring("<script xmlns='http://www.w3.org/1999/xhtml' src='%s' type='text/javascript'  />" % scriptUrl)
```

Comment:
👍 verified that this construction is safe, shenaniganry with single-quotes in `scriptUrl` results in an exception.

Reply:
I think that's pure luck thanks to the use of a self-closing tag, which makes it impossible to inject additional elements.  It's certainly possible to inject additional attributes onto the `<script>` tag, although I'm struggling to imagine what you could do that's worse than being allowed to specify the URL to arbitrary JavaScript :-)

There's a straightforward way to make this safe, which I've done in 115632b.

Reply:
> I'm struggling to imagine what you could do that's worse than being allowed to specify the URL to arbitrary JavaScript :-)

hah good point 🤣